### PR TITLE
Updated Bukkit to 1.20.6

### PR DESCRIPTION
**Pull request by OakLoaf**
In this PR I have also set the Github gradle-build action to run on Java 21 as it is required by 1.20.6.

I am having some trouble getting the gradle dependencies to work so have been unable to test this at all.

### CODE REVIEW

**Comment by OakLoaf** [comment]
If anyone has any ideas as to why these dependencies aren't working that'd be really helpful:
```
A problem occurred configuring project ':platforms:bukkit:nms:v1_20_R6'.
> Failed to notify project evaluation listener.
   > Could not resolve all files for configuration ':platforms:bukkit:nms:v1_20_R6:paperweightDevelopmentBundle'.
      > Could not find io.papermc.paper:dev-bundle:1.20.6-R0.1-20240530.204353-104.
        Searched in the following locations:
          - https://repo.maven.apache.org/maven2/io/papermc/paper/dev-bundle/1.20.6-R0.1-SNAPSHOT/dev-bundle-1.20.6-R0.1-20240530.204353-104.pom
          - https://plugins.gradle.org/m2/io/papermc/paper/dev-bundle/1.20.6-R0.1-SNAPSHOT/dev-bundle-1.20.6-R0.1-20240530.204353-104.pom
          - https://maven.fabricmc.net/io/papermc/paper/dev-bundle/1.20.6-R0.1-SNAPSHOT/dev-bundle-1.20.6-R0.1-20240530.204353-104.pom
          - https://repo.codemc.org/repository/maven-public/io/papermc/paper/dev-bundle/1.20.6-R0.1-SNAPSHOT/dev-bundle-1.20.6-R0.1-20240530.204353-104.pom
          - https://papermc.io/repo/repository/maven-public/io/papermc/paper/dev-bundle/1.20.6-R0.1-SNAPSHOT/dev-bundle-1.20.6-R0.1-20240530.204353-104.pom
          - https://files.minecraftforge.net/maven/io/papermc/paper/dev-bundle/1.20.6-R0.1-SNAPSHOT/dev-bundle-1.20.6-R0.1-20240530.204353-104.pom
          - https://maven.quiltmc.org/repository/release/io/papermc/paper/dev-bundle/1.20.6-R0.1-SNAPSHOT/dev-bundle-1.20.6-R0.1-20240530.204353-104.pom
          - https://jitpack.io/io/papermc/paper/dev-bundle/1.20.6-R0.1-SNAPSHOT/dev-bundle-1.20.6-R0.1-20240530.204353-104.pom
          - https://s01.oss.sonatype.org/content/repositories/snapshots/io/papermc/paper/dev-bundle/1.20.6-R0.1-SNAPSHOT/dev-bundle-1.20.6-R0.1-20240530.204353-104.pom
          - https://repo.papermc.io/repository/maven-public/io/papermc/paper/dev-bundle/1.20.6-R0.1-SNAPSHOT/dev-bundle-1.20.6-R0.1-20240530.204353-104.pom
        Required by:
            project :platforms:bukkit:nms:v1_20_R6

Possible solution:
 - Declare repository providing the artifact, see the documentation at https://docs.gradle.org/current/userguide/declaring_repositories.html
```

**Comment by Arthurmeade12** [comment]
If you're not aware it looks like a lot of work has already been done on this branch here https://github.com/PolyhedralDev/Terra/commits/dev/1.20.6/

**Comment by justaureus** [comment]
> If you're not aware it looks like a lot of work has already been done on this branch here https://github.com/PolyhedralDev/Terra/commits/dev/1.20.6/

Not really no, that's just the initial work required for 1.20.6.  It's also only for Fabric not Paper.

**Comment by OakLoaf** [comment]
> If you're not aware it looks like a lot of work has already been done on this branch here https://github.com/PolyhedralDev/Terra/commits/dev/1.20.6/

This is a pull request for that branch :)

**Comment by IkyMax** [comment]
Hey! This happens because in the paper repo is marked as

        `const val paperBuild = "$minecraft-R0.1-20240530.204353-103"`
        
But you're using

        `const val paperBuild = "$minecraft-R0.1-20240530.204353-104"`
        
this is in [buildSrc/src/main/kotlin/Versions.kt](https://github.com/PolyhedralDev/Terra/compare/master...OakLoaf:Terra:dev/1.20.6/bukkit#diff-c1c2c95e10d73bad89764533c4e4cd48653a02484f07df8316b8fa9fc15da0c8), anyways, you should use the newest devbundle with

`        const val paperBuild = "$minecraft-R0.1-20240601.143523-105"`

**Comment by IkyMax** [comment]
Forget what i Said LMAO

paperBuild and paperDevBundle are two separated repos and they're NOT synced in builds, so in order to use them both, you need to specify the versions for both! Use

        const val paperBuild = "$minecraft-R0.1-20240601.143523-106"

and this

        const val paperDevBundle = "$minecraft-R0.1-20240601.143523-105"

**Comment by OakLoaf** [comment]
> Forget what i Said LMAO
> 
> paperBuild and paperDevBundle are two separated repos and they're NOT synced in builds, so in order to use them both, you need to specify the versions for both! Use
> 
> ```
>     const val paperBuild = "$minecraft-R0.1-20240601.143523-106"
> ```
> 
> and this
> 
> ```
>     const val paperDevBundle = "$minecraft-R0.1-20240601.143523-105"
> ```

Oooh ok, I'll give that a go!

**Comment by IkyMax** [comment]
> Oooh ok, I'll give that a go!

Sure! also, check the java thing lol

**Comment by OakLoaf** [comment]
Yeahh I'll try to work out what's going on there

**Comment by OakLoaf** [comment]
Not having much luck with this one unfortunately.
`:platforms:bukkit:common:main: Could not resolve io.papermc.paper:paper-api:1.20.6-R0.1-20240601.143523-106.`

**Comment by IkyMax** [comment]
weird, let me check your repo lol

**Comment by OakLoaf** [comment]
I'll commit my attempt real quick

**Comment by IkyMax** [comment]
Sorry, i was busy, but i find what you need to change the repo in [DependencyConfig.kt](https://github.com/OakLoaf/Terra/blob/dev/1.20.6/bukkit/buildSrc/src/main/kotlin/DependencyConfig.kt) and in [build.gradle.kts](https://github.com/OakLoaf/Terra/blob/dev/1.20.6/bukkit/buildSrc/build.gradle.kts) from https://papermc.io/repo/repository/maven-public/ to https://repo.papermc.io/repository/maven-public/

Also you need to change the Java version to 21 in [CompilationConfig.kt](https://github.com/OakLoaf/Terra/blob/dev/1.20.6/bukkit/buildSrc/src/main/kotlin/CompilationConfig.kt) but this breaks fabric builds so idk

**Comment by IkyMax** [comment]
I did a bukkit build only with java version 21 and the Paper API error didn't happen but the NMS are broken

**Comment by OakLoaf** [comment]
> Sorry, i was busy, but i find what you need to change the repo in [DependencyConfig.kt](https://github.com/OakLoaf/Terra/blob/dev/1.20.6/bukkit/buildSrc/src/main/kotlin/DependencyConfig.kt) and in [build.gradle.kts](https://github.com/OakLoaf/Terra/blob/dev/1.20.6/bukkit/buildSrc/build.gradle.kts) from https://papermc.io/repo/repository/maven-public/ to https://repo.papermc.io/repository/maven-public/
> 
> Also you need to change the Java version to 21 in [CompilationConfig.kt](https://github.com/OakLoaf/Terra/blob/dev/1.20.6/bukkit/buildSrc/src/main/kotlin/CompilationConfig.kt) but this breaks fabric builds so idk

Ahhh ok, didn't know about the dependency kotlin file, I'll give that one a go.

**Comment by OakLoaf** [comment]
> I did a bukkit build only with java version 21 and the Paper API error didn't happen but the NMS are totally broken, you also would need to migrate the bukkit API to the [Bukkit long-standing API](https://forums.papermc.io/threads/important-dev-psa-future-removal-of-cb-package-relocation.1106/) inside [RegistryFetcher.java](https://github.com/OakLoaf/Terra/blob/17acde09a31d81063a29b877d0bfdd9381e09b68/platforms/bukkit/nms/v1_20_R6/src/main/java/com/dfsek/terra/bukkit/nms/v1_20_R6/RegistryFetcher.java)

I a bit confused, as I did this in my initial commit lol

**Comment by OakLoaf** [comment]
I've tested this with the overworld pack on a 1.20.6 server, everything appears to be working as intended with no errors in console.

**Comment by OakLoaf** [comment]
> builds fine, runs overworld 2.0 fine but i do get some errors and warnings: https://files.catbox.moe/wdw1m4.txt

Is this error from another pack? Based off of the error I would assume you need to change `minecraft:skeleton` to `skeleton`

**Comment by OakLoaf** [comment]
> its https://github.com/PolyhedralDev/TerraOverworldConfig/tree/2.0 and does not show up on 1.20.4 which is why i am mentioning this here

I've just tested on the latest modrinth release on a 1.20.4 server and the issue is still present.

**Comment by IkyMax** [comment]
Hey! this is normal? lol

image

**Comment by OakLoaf** [comment]
Good spot, let me resolve this!

**Comment by OakLoaf** [comment]
That should be resolved in https://github.com/PolyhedralDev/Terra/pull/458/commits/9e60db4b0da03f59c7991a93659b9c4ea40f8362

**Comment by OakLoaf** [comment]
I've updated this pr for Bukkit 1.21, I'm unsure if this reflection here [`4e7de50`](https://github.com/PolyhedralDev/Terra/pull/458/commits/4e7de501c038be60df2f18004de6d7323e5a4282#diff-13d89cf4f0afbeee1bc62e965041efdaf163e55841786cdc2646686323664765R45-R60) is the best way to do this but it's the only thing I could personally see.

*This is currently untested*

**Comment by TobyWarren** [comment]
Doing a test build now. Will try and setup a server to test on.

**Comment by TobyWarren** [comment]
[failure.txt](https://github.com/user-attachments/files/15860392/failure.txt)
Not sure if this is something on my end, but it still doesn't build.

**Comment by IkyMax** [comment]
> [failure.txt](https://github.com/user-attachments/files/15860392/failure.txt) Not sure if this is something on my end, but it still doesn't build.

Have you builded the Spigot version only? this usually happens because you only ran gradle build instead of the specific platform build

**Comment by TobyWarren** [comment]
> > [failure.txt](https://github.com/user-attachments/files/15860392/failure.txt) Not sure if this is something on my end, but it still doesn't build.
> 
> Have you builded the Spigot version only? this usually happens because you only ran gradle build instead of the specific platform build

Whats the command for that again?

**Comment by OakLoaf** [comment]
> Built your branch and paper 1.21, heres the log: https://files.catbox.moe/si1g9m.log

Ahh, seems like cloud will need an update

File: platforms/bukkit/common/src/main/java/com/dfsek/terra/bukkit/util/VersionUtil.java [hunk 1]
```diff
@@ -77,13 +77,13 @@ public boolean isSpigot() {
...
 
-        private static final Pattern VERSION_PATTERN = Pattern.compile("v?(\\d+)_(\\d+)_R(\\d+)");
+        private static final Pattern VERSION_PATTERN = Pattern.compile("(\\d+)\\.(\\d+)\\.(\\d+)");
         private final int major;
         private final int minor;
         private final int patch;
 
         private MinecraftVersionInfo() {
-            this(Bukkit.getServer().getClass().getPackage().getName().split("\\.")[3]);
+            this(Bukkit.getServer().getMinecraftVersion());
```
Comment (PikaMug) [inline platforms/bukkit/common/src/main/java/com/dfsek/terra/bukkit/util/VersionUtil.java:86]: This is a Paper-only method. According to https://terra.polydev.org/install/bukkit.html Spigot is still supported (albeit not recommended).

Perhaps you could do something with `Bukkit.getServer().getBukkitVersion().split("-")[0]`

Comment (OakLoaf) [reply platforms/bukkit/common/src/main/java/com/dfsek/terra/bukkit/util/VersionUtil.java:86]: Ahh I wasn't aware that Spigot was still supported, I'll make that change

**Comment by TobyWarren** [comment]
Looks like there is an issue with Terra pulling version information on 1.21 Paper, have not tested Spigot. Log attached.
[latest.log](https://github.com/user-attachments/files/15878382/latest.log)

**Comment by OakLoaf** [comment]
Reverted 1.21 commits and moved them to https://github.com/PolyhedralDev/Terra/pull/461

**Comment by Zahikusa** [comment]
I really need your plugin for my new server :) Anything I can help with to get this running on minecraft 1.20.6?

**Comment by OakLoaf** [comment]
It should run on 1.20.6 already if you compile it yourself

**Comment by justaureus** [comment]
These commits look good, tested this PR locally and all seems to be well.  Thank you for your contributions!

**Pull request merged**